overlord/state/taskrunner.go

```diff
@@ -84,22 +90,27 @@ func (r *TaskRunner) run(fn HandlerFunc, task *Task) {
 	tomb := &tomb.Tomb{}
 	r.tombs[task.ID()] = tomb
 	tomb.Go(func() error {
-		err := fn(task, tomb)
+		tomb.Kill(fn(task, tomb))
 
 		r.state.Lock()
 		defer r.state.Unlock()
 		halted := task.HaltTasks()
-		if err == nil {
+		switch tomb.Err() {
+		case Retry:
+			// Do nothing. Handler asked to try again later.
+			// TODO: define how to control retry intervals,
+			// right now things will be retried at the next Ensure
+		case nil:
 			task.SetStatus(DoneStatus)
 			if len(halted) > 0 {
```

Comment:
Why isn't this len(task.Halted())?  I don't see halted being used anywhere else.